### PR TITLE
Chat: new settings

### PR DESCRIPTION
**Pull request by nerzhul**
I decided to backport some changes from my inactive but working fork to MT.

I added new features:

* [x] Chat messages length limit
* [x] Message rate limiting
* [x] Message rate kicking

For this change i changed a peer_id to Player* pointer because i will use a player attribute to save the current rate

### CODE REVIEW

**Comment by nerzhul** [comment]
Ready for review @sfan5 @est31 @sofar

**Comment by sofar** [comment]
:+1: on the code alltogether, this is fairly simple and straightforward.

Minor nitpicks:
- cache `g_settings` perhaps?
- the names of the settings are not very consistent. Perhaps name them all `chat_max_message_size`, `chat_message_limit_per_10s` etc?

File: src/player.cpp [hunk 1]
```diff
@@ -229,7 +229,10 @@ void Player::clearHud()
 
 RemotePlayer::RemotePlayer(IGameDef *gamedef, const char *name):
 	Player(gamedef, name),
-	m_sao(NULL)
+	m_sao(NULL),
+	m_last_chat_message_sent(time(NULL)),
```
Comment (sfan5) [inline src/player.cpp]: Shouldn't this use the `TimeGetter` from `porting`?

**Comment by sfan5** [comment]
:+1:

**Comment by nerzhul** [comment]
@sofar will rename to more consistent names, yes, agree with your better naming scheme.
@sfan5 will look at timegetter, didn't remember about it

File: src/player.cpp (hunk 1 above)
Comment (nerzhul) [inline src/player.cpp]: TimeGetter is only available in clientlauncher.h, and i think we don't need this too generic way

**Comment by nerzhul** [comment]
@sofar @sfan5 @est31 i fixed named and add cache settings, is this okay for you ?

**Comment by nerzhul** [comment]
Commit squashed, waiting for build

**Comment by rubenwardy** [comment]
A better message would be `Add chat spam abuse protection` or similar. It's more descriptive

**Pull request merged**